_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,538 @@
+Now, click `Save`.
+Switch to another tab that appears, called `Credentials`.
+Leave `Client Authenticator` set to `Client Id and Secret`, and take note of the `Secret`. We'll need it later.
+
+What we've done here is configure the kafka-broker 'client' in a way that it can directly authenticate with Keycloak in its own name, using its own Client ID and Secret, and at the same time can't be used for browser based login by other users.
+This is a typical 'client' configuration for microservices, regardless of whether they are acting as application clients, or application servers. 
+
+![Create Client in Keycloak Console](/assets/2019-10-23-keycloak.png)
+
+All Kafka brokers will share the same 'client' configuration since they are just different instances of the same service.
+
+Next, we configure a 'client' for each service that acts as a Kafka client. These are Kafka producers, consumers, connectors, streams applications - any application that talks to Kafka brokers. 
+
+In general this is done by following the same steps as for 'kafka-broker', but giving each new client its own `Client ID`, and taking note of the generated `Secret`.
+
+For this demo we'll need two more clients: 'kafka-producer' and 'kafka-consumer'.
+
+When you are done it is wise to export the current state of the realm configuration.
+Since we are running Keycloak as a stateless pod, it will happen sooner rather than later that the pod will be deleted and replaced with a fresh new one, and all state will be gone.
+Exporting the realm gives us a way to restore this state when this happens.
+
+To export the realm click on `Export` in main navigation on the left.
+
+Then in `Partial Export` page turn `Export clients` to ON, and click `Export`.
+Note, that secrets in the exported file get lost, and if you ever perform `Import`, you'll have to use the `Regenerate Secret` button in the `Credentials` tab of each client.
+
+
+
+## Installing Strimzi
+
+Before we can start a Kafka cluster we first have to install the custom resource definitions (CRDs) for Strimzi.
+
+The default installation assumes that the namespace where Kafka cluster is deployed is called 'myproject'.
+However, we will assume here that kafka cluster is running in 'kafka' namespace.
+
+In order to adjust the namespace where CRDs are installed, we can do an on-the-fly transformation of installation file:
+
+    kubectl create ns kafka
+    
+    curl -L https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.14.0/strimzi-cluster-operator-0.14.0.yaml \
+      | sed 's/namespace: .*/namespace: kafka/' \
+      | kubectl apply -f - -n kafka
+
+Make sure the cluster operator is running:
+
+    kubectl -n kafka get pods
+
+    
+NOTE: If you're using RBAC you may first need to grant current user ('developer' by default) cluster-admin role. For example on OpenShift:
```

Comment:
Since you so far talked about Minikube only and some parts (`sudo minikube stunnel`) are Minikube specific, I would probably leave this out.

Reply:
You can also run minikube with RBAC by using: --extra-config=apiserver.Authorization.Mode=RBAC

I'm trying to address some typical hurdles people in different environments might encounter. While it's true that you can't address every different circumstance I'm generally happier as a reader if I can get things working quickly without having to figure it out. Also I like it when it's mentioned that something may not just work, and tells me why.

Reply:
Sorry, I maybe didn't placed the command to the best place. I think mentioning RBAC is fine. But I'm less sure about mentioning OpenShift and GKE.

Reply:
I think you may not have seen the [latest changes](https://github.com/strimzi/strimzi.github.io/pull/105/commits/d29229d0d32af07153eb9b17f08d7bfee65870cc).

As far as mentioning various Kubernetes options and trying to address possible issues with them, I think that if someone already has a working kubectl tied to some cluster be it local or cloudy they would not go and install minikube just to follow this example. They will try use their existing Kubernetes environment, and almost certainly hit this issue. But then ideally I could just link to Strimzi install docs where all the troubleshooting for different Kubernetes environments is also covered. Any suggestions where in our existing docs to point to for that?